{"cells":[{"cell_type":"markdown","metadata":{"id":"195bMV1d0fL9"},"source":["## **Summary Code Note**\n","This file summarizes the codes, functions, and packages that we wrote and ran to retrieve mtDNA sequences from NCBI, and then clean the raw data to get the final dataset 3 (which only includes complete genomes), and finally use the summary information of each sequence to create the Isolate Explanation table. </br>\n","For the purpose of showing only the way to get expected outcomes, and giving the readers a whole picture of section Retrieving Data and Data Wrangling in Method & Materials of an article, we only showed the codes that call the modules with functions of main packages, and some extra codes (loops) repeating the same functions. To know more about the specific details of codes of how the packages, modules, functions, etc., were written, you can look up the git link below:\n","\n","https://github.com/vy-phung/Haplogroup.git"]},{"cell_type":"markdown","metadata":{"id":"AyVUgrJFLvyl"},"source":["Notices:\n","- Most of the files input as well as the saved outputs were stored in the [google drive folder](https://drive.google.com/drive/folders/15OPBImGAG51vukHfADV3-zG8RKU8A9fe?usp=drive_link). </br>\n","The paths of the those input and output files are showed in the codes below sorted in the same order in the google drive above."]},{"cell_type":"markdown","metadata":{"id":"aGv8tuieL07w"},"source":["\n","**CONTENTS:**\n","1. Get Data from NCBI\n","* Entrez Direct\n","* Missing Data\n","2. Data wrangling\n","* Dataset 1\n","* Dataset 2\n","* Dataset 3\n","3. Tables\n","* Isolate Explanation Table\n","* Table 1\n","* Table 2\n","* Table 3 and its subtables"]},{"cell_type":"markdown","metadata":{"id":"Q42i00Uq017m"},"source":["### **1. Get Data from NCBI**"]},{"cell_type":"markdown","metadata":{"id":"1CQ5RaLv5DXN"},"source":["#### **Entrez Direct**"]},{"cell_type":"markdown","metadata":{"id":"-j9r_cYm3ZR4"},"source":["- To retrieve mtDNA sequences of 11 countries in South East Asia, we used [Entrez Direct](https://www.ncbi.nlm.nih.gov/books/NBK179288/) to search for the common keywords: </br>\n","`\"Homo sapiens AND mitochondrion AND <Country Name>`\n","\n","- Below is a bash script file we wrote to get and then save the collected data. The file \"countries.txt\" in DataList contains the names of 11 countries (Brunei, Cambodia, Indonesia, Laos, Malaysia, Myanmar, Philippines, Singapore, Thailand, Timor-Leste, Viet Nam). The collected fasta files were saved in OldCountryFasta folder."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Gl1EQPRM2Nkq"},"outputs":[],"source":["# create a bash script file to run and get mtDNA off 11 countries by using the above mentioned keywords\n","import os\n","os.system('''echo \"#!/bin/bash\n","# Download and Install Entrez Direct\n","sh -c \"$(wget -q https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh -O -)\"\n","export PATH=${HOME}/edirect:${PATH}\n","\n","# run code to get data of 11 countries\n","DataList=/content/drive/MyDrive/OUCRUwork/RetrieveData/others/extraData/countries.txt\n","Field_Separator=$IFS\n","IFS=,\n","for val in `cat $DataList`\n","do ${HOME}/edirect/esearch -db nucleotide -query \"Homo sapiens AND mitochondrion AND $val\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta\n","> /content/drive/MyDrive/RetrieveData/OldCountryFasta/$val.fasta;\n","done \" >> /content/drive/MyDrive/OUCRUwork/RetrieveData/others/codes/NCBIBash.sh''')"]},{"cell_type":"code","source":["def saveFile(name,content):\n","  Name = name\n","  fi = open(Name, \"a\")\n","      # Add new change in to saved file\n","  with open(Name, \"w\") as external_file:\n","    add_text = content\n","    print(add_text, file=external_file)\n","    external_file.close()"],"metadata":{"id":"R2EHIYlPGJKf","executionInfo":{"status":"ok","timestamp":1711495767594,"user_tz":240,"elapsed":9,"user":{"displayName":"Vy Phung","userId":"04206573663256464938"}}},"execution_count":1,"outputs":[]},{"cell_type":"code","source":["file = '''\n","#!/bin/bash\n","\n","greeting () {\n","  echo \"Hello $1\"\n","}\n","\n","#greeting \"Joe\"\n","'''\n","saveFile(\"hello.sh\",file)"],"metadata":{"id":"J8UmvMMJGHzk","executionInfo":{"status":"ok","timestamp":1711496502571,"user_tz":240,"elapsed":138,"user":{"displayName":"Vy Phung","userId":"04206573663256464938"}}},"execution_count":32,"outputs":[]},{"cell_type":"code","source":["! cat hello.sh"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vhxs5OJXGnYP","executionInfo":{"status":"ok","timestamp":1711496507222,"user_tz":240,"elapsed":215,"user":{"displayName":"Vy Phung","userId":"04206573663256464938"}},"outputId":"c1c58aeb-933f-46a5-8c35-b9f6e734cbd3"},"execution_count":33,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","#!/bin/bash\n","\n","greeting () {\n","  echo \"Hello $1\"\n","}\n","\n","#greeting \"Joe\"\n","\n"]}]},{"cell_type":"code","source":["! source hello.sh; greeting \"Jay\""],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"6nFIbAE4Gads","executionInfo":{"status":"ok","timestamp":1711496622482,"user_tz":240,"elapsed":141,"user":{"displayName":"Vy Phung","userId":"04206573663256464938"}},"outputId":"e32b4249-8ea1-4c43-b96f-f1be6ec897e2"},"execution_count":42,"outputs":[{"output_type":"stream","name":"stdout","text":["Hello Jay\n"]}]},{"cell_type":"markdown","metadata":{"id":"DuAovrI75VoA"},"source":["#### **Missing data**\n","\n","After gaining the data from the above common keywords, we ralized that there were still some more missing data if we search the keywords specifically. Therfore, below is the the way we got for each specific country"]},{"cell_type":"markdown","metadata":{"id":"UWpj4t5mBh1y"},"source":["**Myanmar** </br>\n","New Data:\n","- 21219640.Inland post-glacial dispersal in East Asia revealed by mitochondrial haplogroup M9a'b: Myanmar, Vietnam (filter them): keywords for Myanmar: HM346895, HM346896\n","- 24467713.Summerer et al. (2014).txt: Myanmar: missing 327 small coding region JX288765-JX289091 among 371 files of this article (keywords: Large-scale mitochondrial DNA analysis in Southeast Asia reveals evolutionary effects of cultural isolation in the multi-ethnic population of Myanmar)\n","- 25826227.Li et al. (2015).txt: all 937 files for Myanmar but only 92 files exist (keywords: Ancient inland human dispersals from Myanmar into interior East Asia since the Late Pleistocene)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jdH2j81GBvoE"},"outputs":[],"source":["! for i in HM346895 HM346896; do ${HOME}/edirect/esearch -db nucleotide -query $i -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Myanmar/extra/$i.fasta; done\n","\n","! ${HOME}/edirect/esearch -db nucleotide -query \"Large-scale mitochondrial DNA analysis in Southeast Asia reveals evolutionary effects of cultural isolation in the multi-ethnic population of Myanmar\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Myanmar/extra/$i.fasta\n","\n","! ${HOME}/edirect/esearch -db nucleotide -query \"Ancient inland human dispersals from Myanmar into interior East Asia since the Late Pleistocene\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Myanmar/extra/$i.fasta"]},{"cell_type":"markdown","metadata":{"id":"oi_z2gyU3WAI"},"source":["**Philippines** </br>\n","New data:\n","+ 21281460.Loo et al. (2011).txt: all 46 files for Philip but only exist 12 (keyword: Genetic affinities between the Yami tribe people of Orchid Island and the Philippine Islanders of the Batanes archipelago and Philippines)\n","+ 21796613.Scholes et al. (2011).txt: 60 files for Philip but only exist 9 (keywords: Genetic diversity and evidence for population admixture in Batak Negritos from Palawan)\n","+ Philippines: '28535779.Carriers of mitochondrial DNA macrohaplogroup R colonized Eurasia.Larruga et al. (2017)': keywords: \"Carriers of mitochondrial DNA macrohaplogroup R colonized Eurasia AND Philippines\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rlEgMKZMDoAP"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query 'Genetic affinities between the Yami tribe people of Orchid Island and the Philippine Islanders of the Batanes archipelago and Philippines' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Philippines/extra/Philippines2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'Genetic diversity and evidence for population admixture in Batak Negritos from Palawan' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Philippines/extra/Philippines2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'Carriers of mitochondrial DNA macrohaplogroup R colonized Eurasia AND Philippines' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Philippines/extra/Philippines2.fasta"]},{"cell_type":"markdown","metadata":{"id":"IDwv9REX4WE4"},"source":["**Singapore** </br>\n","New data:\n","+ 36382667.Sui et al. (2023).txt: there are 7 and already exist 5 so there are 2 more but no idea if these 2 ref seqs are from Singapore or not (keywords: Death associated protein‑3 (DAP3) and DAP3 binding cell death enhancer‑1 (DELE1) in human colorectal cancer, and their impacts on clinical outcome and chemoresistance)\n","+ 37025097.Zhao et al. (2023).txt: 14 files and already exist 11 files but don’t know about the other 3 belongs to Singapore or not"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GqC7rB3DFVTt"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query \"Death associated protein‑3 (DAP3) and DAP3 binding cell death enhancer‑1 (DELE1) in human colorectal cancer, and their impacts on clinical outcome and chemoresistance\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Singapore/extra/Singapore2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"Effect of COP1 in Promoting the Tumorigenesis of Gastric Cancer by Down-Regulation of CDH18 via PI3K/AKT Signal Pathway\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Singapore/extra/Singapore2.fasta"]},{"cell_type":"markdown","metadata":{"id":"spNQKJDa4rSQ"},"source":["**Thailand** </br>\n","New data:\n","+ 11310578.Mitochondrial DNA polymorphisms in Thailand.Fucharoen et al. (2001)\n","+ 19148289.The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers.Jin et al. (2009): Thai (80, keyword: The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers and Thailand), Vietnam (84, keyword: Viet Nam)\n","+ 27837350.Complete mitochondrial genomes of Thai and Lao populations indicate an ancient origin of Austroasiatic groups and demic diffusion in the spread of Tai–Kadai languages. Kutanan et al. (2017): Thai, Lao (search on the table of isolate name): total 1234\n","+ 32304863.A Matrilineal Genetic Perspective of Hanging Coffin\n","Custom in Southern China and Northern Thailand (its old name is Unpublished.The Population History and Cultural Dispersal Pattern of Hanging.Zhang et al. (2020))\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"J5CJWGWh_xew"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query 'Mitochondrial DNA polymorphisms in Thailand' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Thailand/extra/Thailand2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers AND Thailand ' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Thailand/extra/Thailand2.fasta\n","\n","'''27837350.Complete mitochondrial genomes of Thai and Lao populations indicate an ancient origin of Austroasiatic groups and demic diffusion in the\n","spread of Tai–Kadai languages. Kutanan et al. (2017): Thai, Lao (search on the table of isolate name):\n","total 1234: Laos: LUA101-LUA149: LA1 + VIE101-VIE149: LA2; the others are Thai'''\n","# download data for both Thai and Lao\n","!${HOME}/edirect/esearch -db nucleotide -query \"Complete mitochondrial genomes of Thai and Lao populations indicate an ancient origin of Austroasiatic groups and demic diffusion in the spread of Tai–Kadai languages \" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Laos/extra/Laos_Thai.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"The Population History and Cultural Dispersal Pattern of Hanging\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/Dataset1/Thailand/extra/Thailand.fasta"]},{"cell_type":"markdown","metadata":{"id":"ogigvxd1_gad"},"source":["**Laos** </br>\n","New Data:\n","- 21333001.Southeast Asian diversity: first insights into the complex mtDNA structure of Laos.Bodner et al. (2011): 214"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jBrH5lXf_ljj"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query \"Southeast Asian diversity: first insights into the complex mtDNA structure of Laos\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Laos/extra/Laos_Thai.fasta"]},{"cell_type":"markdown","metadata":{"id":"EaQvWDtD52JL"},"source":["**Timor-Leste** </br>\n","New data:\n","+ Genetic admixture history of Eastern Indonesia as revealed by Y-chromosome and mitochondrial DNA analysis.Mona et al. (2009): No clear location/country but this article includes for countries: Indo, Timor (330 files)\n","+ 25757516.Gomes et al. (2015).txt: this study has 324 files which are all from East Timor (Timor-Leste) (KJ655583-KJ655889: D-loop, KJ676774-KJ676790: complete genome) (keywords: Human settlement history between Sunda and Sahul: a focus on East Timor (Timor-Leste) and the Pleistocenic mtDNA diversity)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LcAgTv77Fq4y"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query 'Genetic admixture history of Eastern Indonesia as revealed by Y-chromosome and mitochondrial DNA analysis' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Timor-Leste/extra/Timor2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'Human settlement history between Sunda and Sahul: a focus on East Timor (Timor-Leste) and the Pleistocenic mtDNA diversity' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Timor-Leste/extra/Timor2.fasta"]},{"cell_type":"markdown","metadata":{"id":"kAVgVjMh5_Qm"},"source":["**Viet Nam** </br>\n","New data:\n","+ 21219640.Inland post-glacial dispersal in East Asia revealed by mitochondrial haplogroup M9a'b: Myanmar, Vietnam: keywords for VN: HM346881, HM346883, HM346885, HM346886, HM346889\n","+ .Direct Submission.VN.Phan et al. (2016): Vietnam (DQ834255, DQ834258)\n","+ 19148289.The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers.Jin et al. (2009): Vietnam (84, keyword: The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers and Viet Nam)\n","+ 20513740.Tracing the Austronesian footprint in Mainland Southeast Asia: a perspective from mitochondrial DNA.Peng et al. (2010): 335 (Cham+Kinh)\n","+ '.Direct Submission.Phan et al. (2016)': there are 10 files for VN if search by keyword “Phan (2016) AND Homo sapiens AND mitochondrion” and among them already existed 2 files. They dont have a title for this so I still cannot find the article and there is no explanation for the isolate; the isolate only has “VN”"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"k5oninHeGTi7"},"outputs":[],"source":["! for i in HM346881 HM346883 HM346885 HM346886 HM346889; do ${HOME}/edirect/esearch -db nucleotide -query $i -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Vietnam/extra/$i.fasta; done\n","\n","! for i in DQ834255 DQ834258; do ${HOME}/edirect/esearch -db nucleotide -query $i -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Vietnam/extra/$i.fasta; done\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'The Peopling of Korea Revealed by Analyses of Mitochondrial DNA and Y-Chromosomal Markers AND Viet Nam' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Vietnam/extra/Vietnam2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query 'Tracing the Austronesian footprint in Mainland Southeast Asia: a perspective from mitochondrial DNA' -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Vietnam/extra/Vietnam2.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"Phan (2016) AND Homo sapiens AND mitochondrion\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/Dataset1/Vietnam/extra/Vietnam.fasta"]},{"cell_type":"markdown","metadata":{"id":"0mvGugR_6geN"},"source":["**Malaysia** </br>\n","1. Search term \"Single, rapid coastal settlement of Asia revealed by analysis of complete mitochondrial genomes AND Malaysia \":\n","- Miss data for Malaysia which is 267 files but all are control region\n","2. Search term \"Single, rapid coastal settlement of Asia revealed by analysis of complete mitochondrial genomes AND Malay\":\n","- Miss 4 complete genome of Malaysia: </br>\n","2: 9_N21(Tor57), 10_M21c(Tor61): Aboriginal Malay (Semelai) (using key word Malay) </br>\n","2: 7_N22(Tor55), 12_M22(Tor63): Aboriginal Malay (Temuan) (using key word Malayu)\n","3. 16982817.Hill et al. (2006): 6 files and one of them which is ORA131B already exist but the other not (keywords: Phylogeography and ethnogenesis of aboriginal Southeast Asians AND Malaysia)\n","4. 22729749.Evolutionary history of continental southeast asians: 'early train' hypothesis based on genetic analysis of mitochondrial and autosomal\n","DNA data.Jinam et al. (2012): Malay: 86 genome: 23 Bidayuh (BD); 24 Jehai (JH); 21 Seletar (SL); 18 Temuan (TM)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0Z9EV3p06lHm"},"outputs":[],"source":["# add more 267 files having control region for Malaysia\n","!${HOME}/edirect/esearch -db nucleotide -query \"Single, rapid coastal settlement of Asia revealed by analysis of complete mitochondrial genomes AND Malaysia \" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/others/extraData/Malaysia_extra.fasta\n","\n","# add more files for complete genome of Malaysia (using key word Malay) (3 files)\n","!${HOME}/edirect/esearch -db nucleotide -query \"Single, rapid coastal settlement of Asia revealed by analysis of complete mitochondrial genomes AND Malay \" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/others/extraData/Malaysia_extra_complete.fasta\n","\n","# using key word Malayu (2 files)\n","!${HOME}/edirect/esearch -db nucleotide -query \"Single, rapid coastal settlement of Asia revealed by analysis of complete mitochondrial genomes AND Malayu \" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta >> /content/drive/MyDrive/RetrieveData/others/extraData/Malaysia_extra_complete.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"Phylogeography and ethnogenesis of aboriginal Southeast Asians AND Malaysia\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Malaysia/extra/Malaysia3.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"Evolutionary history of continental southeast asians: 'early train' hypothesis based on genetic analysis of mitochondrial and autosomal DNA data \" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Malaysia/extra/Malaysia3.fasta"]},{"cell_type":"markdown","metadata":{"id":"Wmbl5nhf7F_z"},"source":["**Indonesia** </br>\n","New Data:\n","- '21407194.Larger mitochondrial DNA than\n","Y-chromosome differences between.Gunnarsdottir et al. (2011)'. Key word: Larger mitochondrial DNA than Y-chromosome differences between matrilocal and patrilocal groups from Sumatra (72 files)\n","- 21407194.Gunnarsdottir et al. (2011): HM596654\n","- 16982817.Hill et al. (2006): 97 files and 4 of them exist (DQ981465-68), the others not (keyword: Phylogeography and ethnogenesis of aboriginal Southeast Asians AND Indonesia)\n","- Unpublished.Ngili et al. (2009).txt: all 206 files from Indonesia (check lại có duplicate với old data ko)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"awwtFxJJ7LsK"},"outputs":[],"source":["!${HOME}/edirect/esearch -db nucleotide -query \"Larger mitochondrial DNA than Y-chromosome differences between matrilocal and patrilocal groups from Sumatra\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/Dataset1/Indonesia/extra/Indonesia.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"HM596654\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/Dataset1/Indonesia/extra/Indonesia.fasta\n","\n","!${HOME}/edirect/esearch -db nucleotide -query \"Phylogeography and ethnogenesis of aboriginal Southeast Asians AND Indonesia\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/RetrieveData/Dataset1/Indonesia/extra/Indonesia.fasta\n","\n","!${HOME}/edirect/esearch -db nuccore -query \"Ngili (2009)\" -sort \"Date Released\" | ${HOME}/edirect/efetch -format fasta > /content/drive/MyDrive/Database/Dataset1/Indonesia/extra/Indonesia3.fasta"]},{"cell_type":"markdown","metadata":{"id":"plQdwLQAHTW0"},"source":["### **2. Data wrangling**\n","\n","From this Data Wrangling, we used the codes we wrote to get the output. The packages containing specific functions were cloned from the [git link](https://github.com/vy-phung/Haplogroup.git) mentioned at first. </br>\n","After cloning, the packages were called and imported to run the function."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wOsmCezjLvyv"},"outputs":[],"source":["# git clone\n","! git clone https://github.com/vy-phung/Haplogroup.git"]},{"cell_type":"markdown","metadata":{"id":"JexAHlZJIoRF"},"source":["#### **Dataset1**\n","After getting raw above data, we created function `SplitSeq` to create the folder Dataset1. </br>\n","This folder includes only unique data which means the duplicated mtDNA sequences getting from NCBI above were removed. </br>\n","Dataset 1 still has reference sequences, D-loop, non-homosapien, etc"]},{"cell_type":"markdown","metadata":{"id":"OmyXVcH6-YZr"},"source":["**Instruction of SplitSeq**</br>\n","**Code:** </br>\n","`SplitSeq(nameFile, country,seqFolder,seqNameFolder,newCountryFasta, exist)`\n","\n","  **Explanation:**\n","  - **nameFile:** input file\n","  - **country:** the country of the input file to run\n","  - **seqFolder:** name of folder of the output fasta files after running and splitting files\n","  - **seqNameFolder:** name of another folder which saves the others besides the above output. For example, after running ReadSummary, this will hold the results of reference aritlces such as authors, titles, pubmedID, etc.\n","  - **newCountryFasta**: a name of folder that we saved a big new fasta file which has the same collected sequences from NCBI, but the difference is they were labellee by the order \"Country.Isolate.AccessionNumber.Haplogroup\"\n","  - **exist:** Only receiving \"Yes\" or \"No\". \"Yes\" means some sequences in a big input fasta file had been processed or already existed, and therefore the function only runs the left over ones. \"No\" means there is no sequences of the input file existed or processed."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8380,"status":"ok","timestamp":1692531013060,"user":{"displayName":"Vy Phung","userId":"04206573663256464938"},"user_tz":-420},"id":"Xpd3_Q6NtIv6","outputId":"d866e2e8-b09a-4abd-86bb-1af7f8d8a825"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","Entrez Direct has been successfully downloaded and installed.\n","\n","In order to complete the configuration process, please execute the following:\n","\n","  echo \"export PATH=/root/edirect:\\${PATH}\" >> ${HOME}/.bashrc\n","\n","or manually edit the PATH variable assignment in your .bashrc file.\n","\n","Would you like to do that automatically now? [y/N]\n","y\n","OK, done.\n","\n","To activate EDirect for this terminal session, please execute the following:\n","\n","export PATH=${HOME}/edirect:${PATH}\n","\n"]}],"source":["# Download and Install Entrez Direct\n","!sh -c \"$(wget -q https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh -O -)\"\n","!export PATH=${HOME}/edirect:${PATH}"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MZxeN9jfo3Ry"},"outputs":[],"source":["# Download haplogrep\n","!curl -sL haplogrep.now.sh | bash"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JrTlpEkxJA4p"},"outputs":[],"source":["# calling function\n","from DataWrangling import Dataset1, openFile\n","# running function to get 11 countries data\n","for country in openFile(\"/content/drive/MyDrive/OUCRUwork/RetrieveData/others/extraData/countries.txt\").split(','):\n","    nameFile = '/content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset1/'+ country +'/old/'+country+'.fasta'\n","    seqFolder = '/content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset1/' + country + '/fasta/'\n","    seqNameFolder = '/content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset1/' + country + '/'\n","    newCountryFasta = '/content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset1/' + country + '/new/'\n","    exist = 'No'\n","    Dataset1.splitSeq(nameFile, country, seqFolder,seqNameFolder,newCountryFasta, exist)"]},{"cell_type":"markdown","metadata":{"id":"JzKx8UU7Lvyw"},"source":["After running `splitSeq` funcion, there are 3 main outputs:\n","+ Saving each new fasta file of a sequence of an input country into the declared folder with its new label name in the order \"Country.Isolate.AccessionNumber.Haplogroup\".\n","+ Additional files seqName.txt and refName.txt helped to read the summary information of the sequences and saved them for the upcoming purpose (creating Isolate Explantion Table, and checking if there are any duplicated sequences).\n","+ A big fasta file including all the sequences of that input country. This file contains the exact same contents as a big old fasta file retrieved from NCBI, but it sequences were labeled with a new name in the order \"Country.Isolate.AccessionNumber.Haplogroup\"."]},{"cell_type":"markdown","metadata":{"id":"lmmbGm1lJrAN"},"source":["#### **Dataset2**\n","\n","After having Dataset1, we ran function `remove(CountryFolder, country)` to remove reference, non-homosapiens sequences.\n","\n","**Explanation:**\n","\n","- **CountryFolder**: the name of the folder of that country saving the output </br>\n","- **country**: name of the input country"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"emJybFyx1WNK"},"outputs":[],"source":["# calling function\n","from DataWrangling.Dataset2_3 import remove\n","from DataWrangling import openFile\n","# run function\n","for country in openFile(\"/content/drive/MyDrive/OUCRUwork/RetrieveData/others/extraData/countries.txt\").split(','):\n","    CountryFolder = '/content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset2/' + country + '/fasta/'\n","    remove(CountryFolder, country)"]},{"cell_type":"markdown","metadata":{"id":"T-4zTOIjLvyw"},"source":["The final output is the sequences which were not reference, non-homosapiens, and D-loop. All the sequences were saved in the new folder named Dataset 2."]},{"cell_type":"markdown","metadata":{"id":"pFga2RRgKTpg"},"source":["#### **Datatset3**\n","After getting Dataset2, we ran function `removeControlReg(country)` to remove control region (or D-loop) of Dataset 2, and remain only complete genomes in this dataset folder.\n","\n","**Explanation:**\n","- **country**: input country"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yoEP9NWt2gk7"},"outputs":[],"source":["# calling function\n","from DataWrangling.Dataset2_3 import removeControlReg\n","from DataWrangling import openFile\n","# run function\n","for country in openFile(\"/content/drive/MyDrive/OUCRUwork/RetrieveData/others/extraData/countries.txt\").split(','):\n","    removeControlReg(country)"]},{"cell_type":"markdown","metadata":{"id":"294v1OzqLvy1"},"source":["The final output is the sequences which were not control region, and only complete genomes. All the sequences were saved in the new folder named Dataset 3 which was also the clean data used for all the data analysis processes and the results of this study."]},{"cell_type":"markdown","metadata":{"id":"cUW9g13aRZh-"},"source":["### **3. Tables**"]},{"cell_type":"markdown","metadata":{"id":"St6xEInbU2Ug"},"source":["In folder RetrieveData/others, there is a file named translation.txt which contains the explanation of all isolates of 4932 sequences to help running the below functions for creating the output tables."]},{"cell_type":"markdown","metadata":{"id":"AbJmgAMTeNH2"},"source":["#### **Isolate Explanation Table**"]},{"cell_type":"markdown","metadata":{"id":"KS2HEa0CLvy3"},"source":["To understand briefly how did we create the Full Isolate Explanation Table, these are the steps to get this table:\n","1. Running the first cell box below to create an orignal Isolate Explanation Table\n","2. We used the file \"Changed_SEA_haplogroups.xlsx\" (to know more how we created this file, scrolling to Table1 section below) and added the columns (References, Ethnicity, Location, Language, Language Family, Haplogroup) having the same AccessionNumber between this file and original Isolate Explanation Table.\n","3. We used the columns References to create more new columns: PubmedID, Title, Years, Author(s), and we add the column \"names\" which were the names we used to label for each sequence\n","4. From column \"Haplogroup\", we ran the code at the third cell box below to add columns Haplogroup1, Haplogroup2, Haplogroup3.\n","5. We ran the code at second cell box below to add polymorphism column.\n","6. We add a column \"Ancient and Present\" by looking for keywords in each article of column \"References\" and classified if their published sequences were ancient or present day."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"1NWA2_dyevEQ"},"outputs":[],"source":["# run function\n","'''Creating an original Isolate Explanation Table which only includes 4 columns:\n","Country, Isolate, AccessionNumber, and Explanation.'''\n","from Tables.IsoTab import *\n","import pandas as pd\n","countries = 'Brunei Cambodia Indonesia Laos Malaysia Myanmar Philippines Singapore Thailand Timor-Leste Vietnam'\n","output = ''\n","for country in countries.split():\n","  table = explainIso(country)\n","  df = IsoTab(table,country)\n","  if len(output) < 1:\n","    output =df\n","  else:\n","    output = pd.concat([output,df])\n","  print('finish',country)\n","output.to_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/IsolateExplanation.xlsx')"]},{"cell_type":"markdown","metadata":{"id":"iAb3P1foAwRP"},"source":["Adding polymorphism"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"UECTwzO8AzLN"},"outputs":[],"source":["'''After having above Isolate explanation table, we wanted to add a column poymorphism for each sequence.\n","Therefore, this function below was used to create that column\n","The input here contains all the fasta files of 11 countries in Dataset3'''\n","\n","# call function\n","from Tables.IsoTab import polymorphism\n","# create input\n","! ls /content/drive/MyDrive/OUCRUwork/RetrieveData/Dataset3/*/fasta/* > Inputs.txt\n","# run function\n","polymorphism.polymorphism('Inputs.txt')\n","polymorphism.tablePoly('Inputs.txt')"]},{"cell_type":"markdown","metadata":{"id":"NAffHzmzLvy4"},"source":["Adding columns Haplo1, Haplo2, Haplo3 from using the original Haplogroup column"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"BppXsmQ8Lvy4"},"outputs":[],"source":["from Tables.IsoTab import polymorphism\n","table3a = pd.read_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table3/table3a_CountryAndEthnicity.xlsx', index_col='Country and Ethnicity')\n","table = pd.concat([table3a.iloc[[0,1],:],table3a.iloc[2:,:].astype(float)])\n","haplo = polymorphism.groupHaplo(table)"]},{"cell_type":"markdown","metadata":{"id":"5RQ7xMEqfVH2"},"source":["#### **Table1**"]},{"cell_type":"markdown","metadata":{"id":"X0vBzWGKLvy4"},"source":["Running the cell box below at this Table1 section created 2 outputs\n","1. Eleven tables for eleven SEA countries which in each table having the columns:  \n","Country, References, Explanation, Ethnicity, Location, Population (or Label), Language, Sample size, Haplogroup's columns (each column is the name of a different haplogroup)\n","2. A table named \"SEA_haplogroups\" which includes the same above columns, but it combined all the 11 tables above into 1 big file.\n","\n","After having \"SEA_Haplogroups\" table, we added a new column \"Language Family\" by using the column \"Ethnicity\" to look up on website World Ethnic Group. We also deleted the column Population (or Label) due to its unnecessary. Moreover, in columns Ethnicity, Location, Language, we added more new information after reading carefully again the papers of the sequences, and included more the nearby location or close-related ethnicity or language for the uncertain information. </br>\n","All these changes were saved in the new file named **\"Changed_SEA_Haplogroups\"**."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"V6mhLOgffXfw"},"outputs":[],"source":["# call function\n","from Tables.tables import table1\n","# run function\n","countries = 'Brunei Cambodia Indonesia Laos Malaysia Myanmar Philippines Singapore Thailand Timor-Leste Vietnam'\n","for country in countries.split():\n","  data = table1(country)\n","  data.to_csv('/content/drive/MyDrive/RetrieveData/tables/table1/'+country+'1.csv')\n","  print(country,'finish')\n","# merge all 11 countries\n","countries = 'Cambodia Indonesia Laos Malaysia Myanmar Philippines Singapore Thailand Timor-Leste Vietnam'\n","df = pd.read_csv('/content/drive/MyDrive/RetrieveData/tables/table1/Brunei1.csv')\n","for country in countries.split():\n","  df1 = pd.read_csv('/content/drive/MyDrive/RetrieveData/tables/table1/'+country+'1.csv')\n","  df = pd.concat([df, df1], ignore_index=True, sort=False)\n","# save the total 11 countries files\n","df = df.fillna(0)\n","df = df.replace(0,'-')\n","df.to_csv('/content/drive/MyDrive/RetrieveData/tables/SEA_haplogroups.csv')"]},{"cell_type":"markdown","metadata":{"id":"xZ2Jpe21ftga"},"source":["#### **Table2**"]},{"cell_type":"markdown","metadata":{"id":"mz4ubQaLLvy5"},"source":["There 2 main outputs of section Table 2 after running the cell box below:\n","1. Eleven haplogroup frequence tables for eleven SEA countries which for each country, the number of haplogroups was different and somes had the same or different haplogroups.\n","2. A big Haplofrequence table includes the total of the haplogroups of all 11 countries.\n","\n","All the tables above have the same columns:\n","- Haplogroup (name of haplogroup)\n","- Number of Individuals (count the appearance of that haplogroup)\n","- Frequency (the percentage of that haplogroup in the total of the number of all haplogroups in that country for first output and in 11 countries for second output)."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"EqDDpltnf82P"},"outputs":[],"source":["# call function\n","from Tables.tables import table2\n","# run function\n","table2 = table2('/content/drive/MyDrive/RetrieveData/tables/table2/Haplogroups.csv')\n","table2.to_csv('/content/drive/MyDrive/RetrieveData/tables/table2/Haplofrequency.csv')"]},{"cell_type":"markdown","metadata":{"id":"GIqywxTSgCvV"},"source":["#### **Table3**\n","\n","In this Table3 section we created 4 main outputs:\n","1. table3a_CountryAndEthnicity:\n","> - Haplogroup column (names of Haplogroup)\n","> - Ethnicities of 11 countries which each country was highlighted in different colors. </br>\n","> - In each country, there were : </br>\n",">> - A Total column (the total numbers of that specific halogroup / the total number of all types of haplogroups in that country). </br>\n",">> - Next to Total column is the name of each different ethnicity appearing in that country (the total number of the specific haplogroup in the specific ethnicity/the total number of all types of haplogroups in that specific ethnicity).\n","2. table3b_Ethnicity:\n","> - Haplogroup column (names of Haplogroup)\n","> - A Total column (the total numbers of that specific haplogroup / the total number of all types of haplogroups). </br>\n","> - Next to Total column is the name of each different ethnicity appearing in all 11 countries (the total number of the specific haplogroup in the specific ethnicity/the total number of all types of haplogroups in that specific ethnicity).\n","3. table3c_LanguageFamily: </br>\n","has the same format of columns as table3b_Ethnicity, but the change the independent variable from Ethnicity to Language Family.\n","4. table3d_CountryAndLanguageFamily: </br>\n","has the same format of columns as table3a_CountryAndEthnicity, but the change the independent variable from Ethnicity to Language Family."]},{"cell_type":"markdown","metadata":{"id":"QTE5EhOwgiUo"},"source":["table 3a_CountryAndEthnicity"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LLdoFGULLvy6"},"outputs":[],"source":["# call function\n","from Tables.tables.table3.table3 import createTable3ad\n","# run function for Ethnicity\n","countries = 'Brunei Cambodia Indonesia Laos Malaysia Myanmar Philippines Singapore Thailand Timor-Leste Vietnam'\n","data = ''\n","groups = ['Country','Ethnicity']\n","for country in countries.split():\n","  df = createTable3ad(country,groups,'/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table1/Changed_SEA_haplogroups.csv',groups[-1])\n","  if len(data) < 1:\n","    data = df\n","  else:\n","    add = df.drop(['Haplogroup'],axis=1)\n","    data = pd.concat([data,add],axis=1)\n","  print(country,'finish')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"D3_Nb0YvLvy6"},"outputs":[],"source":["data.to_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table3/table3a_CountryAndEthnicity.xlsx')"]},{"cell_type":"markdown","metadata":{"id":"QzBERiG4Lvy6"},"source":["table3d_CountryAndLanguageFamily"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GdUZrl2U-Xk7"},"outputs":[],"source":["# call function\n","from Tables.tables.table3.table3 import createTable3ad\n","# run function for Language Family\n","countries = 'Brunei Cambodia Indonesia Laos Malaysia Myanmar Philippines Singapore Thailand Timor-Leste Vietnam'\n","data = ''\n","groups = ['Country','Language family']\n","for country in countries.split():\n","  df = createTable3ad(country,groups,'/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table1/Changed_SEA_haplogroups.csv',groups[-1])\n","  if len(data) < 1:\n","    data = df\n","  else:\n","    add = df.drop(['Haplogroup'],axis=1)\n","    data = pd.concat([data,add],axis=1)\n","  print(country,'finish')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xb50pFZi-aP5"},"outputs":[],"source":["data.to_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table3/table3d_CountryAndLanguageFamily.xlsx')"]},{"cell_type":"markdown","metadata":{"id":"_0eF3iRB-dbA"},"source":["Table 3b,c"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vQmi3iOl-ki8"},"outputs":[],"source":["# call function\n","from Tables.tables.table3.table3 import createTable3bc\n","# run function to get table3b_Ethnicity\n","groups = ['Ethnicity']\n","data = createTable3bc(groups,'/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table1/Changed_SEA_haplogroups.csv','Ethnicity')\n","data.to_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table3/table3b.xlsx')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gmmtQdK6-qlG"},"outputs":[],"source":["# run function to get table3c_LanguageFamily\n","from Tables.tables.table3.table3 import createTable3bc\n","groups = ['Language family']\n","data = createTable3bc(groups,'/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table1/Changed_SEA_haplogroups.csv','Language family')\n","data.to_excel('/content/drive/MyDrive/OUCRUwork/RetrieveData/tables/table3/table3c.xlsx')"]}],"metadata":{"colab":{"provenance":[{"file_id":"1ywFT4Use5YV01DWwgfPfDBjKpKEkywnF","timestamp":1711235568870},{"file_id":"1msJU2e1_WIE51qxN2chEkZU7s2RP1D3A","timestamp":1708839068929},{"file_id":"1j3ujau-e-ssyIQirZDn-VibN-LAxlRZl","timestamp":1690223736992},{"file_id":"1p3xgjdP-wgJbU0AKjPW-bbnlrE3Elfs0","timestamp":1690223570860}]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}